### PR TITLE
kubelet: make registry qps/Burst to limit parallerel pulling count

### LINKED ISSUE #112044: QPS and Burst in kubelet are not being followed
### What happened?

We ran some tests on Burst downloading images on the cluster and came to the conclusion that the kubelet rules not being followed (possibly qps is as well). It appears that the kubelet will just send the requests to the cri on the node without checking the status of what's downloading when in .

```
# journalctl -u containerd | grep PullImage
...
Aug 25 19:57:51 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:57:51.548107884Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.13.3\""
Aug 25 19:57:51 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:57:51.858579104Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.16.10\""
Aug 25 19:57:52 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:57:52.076216882Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.13.5\""
Aug 25 19:57:52 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:57:52.562109721Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.16.7\""
Aug 25 19:57:52 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:57:52.961493900Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.17.1\""
Aug 25 19:57:53 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:57:53.068567947Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.17.2\""
Aug 25 19:57:53 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:57:53.401893433Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.16.5\""
Aug 25 19:57:53 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:57:53.640637236Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.17.4\""
Aug 25 19:57:54 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:57:54.328704856Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.17.5\""
Aug 25 19:57:54 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:57:54.615612963Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.16.8\""
Aug 25 19:57:54 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:57:54.732854159Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.17.6\""
Aug 25 19:57:55 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:57:55.045848045Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.17.3\""
Aug 25 19:57:55 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:57:55.188647607Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.20.0\""
Aug 25 19:57:55 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:57:55.428253785Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.19.0\""
Aug 25 19:57:57 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:57:57.564642386Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.22.2\""
Aug 25 19:57:57 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:57:57.610151248Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.19.1\""
Aug 25 19:57:57 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:57:57.726531931Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.20.5\""
Aug 25 19:57:57 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:57:57.790570825Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.21.5\""
Aug 25 19:57:58 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:57:58.179202573Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.19.3\""
Aug 25 19:57:58 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:57:58.182130444Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.19.2\""
Aug 25 19:57:58 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:57:58.337585791Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.19.4\""
Aug 25 19:58:16 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:58:16.285697752Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.16.10\" returns image reference \"sha256:95e45e68a1373310ea2dd0199681a3545b21e34aa06b0d67623fb5e08dec368f\""
Aug 25 19:58:16 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:58:16.761914592Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.16.7\" returns image reference \"sha256:a356a8707c644c62072e14ec6eb0e359ad08881d7da6fee3a7ae158b92f96fe3\""
Aug 25 19:58:17 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:58:17.093496301Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.17.1\" returns image reference \"sha256:183b53858d7da2c60d5a6cc1314d0d0bb8e1e3b330e20933aa2bdbcf4f1e8d4e\""
Aug 25 19:58:17 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:58:17.094403671Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.20.0\" returns image reference \"sha256:a4e48936364db78edb98edd42166210eeb9ab253341c73580abd6eeaee4c6260\""
Aug 25 19:58:17 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:58:17.357781195Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.16.5\" returns image reference \"sha256:c1fa37765208c82be03a2265e6aa3fec1dfbf22b52aea28278c2e49988b447de\""
Aug 25 19:58:17 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:58:17.454178978Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.17.5\" returns image reference \"sha256:6dd3c4db29bcf14373c711a1ff7e6003fe42c36158476f2539668d384cd7920d\""
Aug 25 19:58:17 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:58:17.705338004Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.17.2\" returns image reference \"sha256:c56dde8f4b257101d0dcce9f3a1b91cfb1ef50af6c86169b3852fb1596df5a2c\""
Aug 25 19:58:17 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:58:17.707619365Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.17.6\" returns image reference \"sha256:b322e6699b6528ba6b3e991f5d38a580fbae2d8b6053408887756c88acb86fb7\""
Aug 25 19:58:17 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:58:17.708168107Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.17.3\" returns image reference \"sha256:bd6b497193c5774c61f8041cf1a89a0f7d47ae189527d9bda792f827d77bad1d\""
Aug 25 19:58:18 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:58:18.105598499Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.19.0\" returns image reference \"sha256:d3278d5a2bbaf44b8f802a13fad52a5a6483b10e4ec9bdcf1196174bb0bde9f5\""
Aug 25 19:58:18 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:58:18.265636047Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.16.8\" returns image reference \"sha256:dca8b0edd3762d817c03308ba46ae9d9d57c970285d4295309e6b970098f7735\""
Aug 25 19:58:18 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:58:18.267775842Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.17.4\" returns image reference \"sha256:d68c03b97310655a949a9e16fc5150d319a826baf88139ff22c263bafd099d22\""
Aug 25 19:58:18 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:58:18.269153608Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.22.2\" returns image reference \"sha256:ef5e13a25c96a230b72fd33701d85769d3068d4857a6980c1c9698ce014729ff\""
Aug 25 19:58:22 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:58:22.056389442Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.19.3\" returns image reference \"sha256:ce0f7b2a8f9c50f42f4ddb4634c02bd5dacbc8bc265766ebb2c245074475e604\""
Aug 25 19:58:22 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:58:22.057494532Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.20.5\" returns image reference \"sha256:e74c36616a2e6cec8d94daa083c638990047e79e6a209f3724e20c050688a438\""
Aug 25 19:58:22 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:58:22.058130036Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.21.5\" returns image reference \"sha256:16dc7c73b399738253243a186500f18e693b25a03acd9be1434df90f6b7cef4a\""
Aug 25 19:58:22 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:58:22.060274844Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.19.4\" returns image reference \"sha256:30d4eb68dcf338c728207535eef37ba6693babf94a6af196344fc43876a13bad\""
Aug 25 19:58:22 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:58:22.060890216Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.19.2\" returns image reference \"sha256:f1a2242d630998581d422de4778363828006f979b0de3b8771dd2d8d18e65292\""
Aug 25 19:58:22 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:58:22.061491581Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.19.1\" returns image reference \"sha256:9b39fbaac9e9daa046f54b6662997536b4b924492552e5d2b09018be36a4454c\""
Aug 25 19:58:22 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:58:22.307424660Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.13.3\" returns image reference \"sha256:3efc460414d9c653856724597620c005190df0c42472981fbd88612647a1d2de\""
Aug 25 19:58:23 test-cc38euv20td4f4nskdjg-standard125-default-0000033e containerd[6825]: time="2022-08-25T19:58:23.144506033Z" level=info msg="PullImage \"us.icr.io/armada-master/calico/node:v3.13.5\" returns image reference \"sha256:e4430bdf613aaf12cff562a9a4e1a70d0191c4e2a3bada325169fa24fbad0ddf\""

```

### What did you expect to happen?

I expected at most 5 images to start downloading and not start another download until at least 1 is finished.

### How can we reproduce it (as minimally and precisely as possible)?

Turn on parallel image pulls on the kubelet and set the qps and burst.
```
serializeImagePulls: false
registryPullQPS: 5
registryBurst: 5
```

Here is the little copy and pastable script I used to test this
```

NODE_NAME=$(kubectl get node --no-headers | head -1 | awk '{print $1}')

# Can be any list of images, these are in our public registry
images=("us.icr.io/armada-master/calico/node:v3.13.3" "us.icr.io/armada-master/calico/node:v3.13.5" "us.icr.io/armada-master/calico/node:v3.16.10" "us.icr.io/armada-master/calico/node:v3.16.5" "us.icr.io/armada-master/calico/node:v3.16.7" "us.icr.io/armada-master/calico/node:v3.16.8" "us.icr.io/armada-master/calico/node:v3.17.1" "us.icr.io/armada-master/calico/node:v3.17.2" "us.icr.io/armada-master/calico/node:v3.17.3" "us.icr.io/armada-master/calico/node:v3.17.4" "us.icr.io/armada-master/calico/node:v3.17.5" "us.icr.io/armada-master/calico/node:v3.17.6" "us.icr.io/armada-master/calico/node:v3.19.0" "us.icr.io/armada-master/calico/node:v3.19.1" "us.icr.io/armada-master/calico/node:v3.19.2" "us.icr.io/armada-master/calico/node:v3.19.3" "us.icr.io/armada-master/calico/node:v3.19.4" "us.icr.io/armada-master/calico/node:v3.20.0" "us.icr.io/armada-master/calico/node:v3.20.5" "us.icr.io/armada-master/calico/node:v3.21.5" "us.icr.io/armada-master/calico/node:v3.22.2")
yamlFile="${HOME}/git/file8"
echo "" > ${yamlFile}
pull_policy="Always"
for image in "${images[@]}"; do

# This awk is mac friendly, I'm not sure if it's linux friendly
nameFriendly=$(echo ${image} | awk -F '/\|:\|\\\.' '{print $7$8$9}')

cat <<EOF >>${yamlFile}
---
apiVersion: v1
kind: Pod
metadata:
  labels:
    run: ${nameFriendly}
  name: ${nameFriendly}
  namespace: default
spec:
  nodeName: ${NODE_NAME}
  containers:
  - image: ${image}
    imagePullPolicy: ${pull_policy}
    name: ${nameFriendly}
    command: ["/bin/sh", "-c", "sleep 10000"]

EOF
done
kubectl apply -f ${yamlFile}
```

Then ssh on to the node and look at the logs of the downloads. In our case, containerd is captured with `journalctl -u containerd`

### Anything else we need to know?

This is using containerd cri.

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.0", GitCommit:"cb303e613a121a29364f75cc67d3d580833a7479", GitTreeState:"clean", BuildDate:"2021-04-08T16:31:21Z", GoVersion:"go1.16.1", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"25+", GitVersion:"v1.25.0-rc.1+IKS", GitCommit:"c18d648ff9e8c02ccccae835394aa53761f7716f", GitTreeState:"clean", BuildDate:"2022-08-17T12:34:27Z", GoVersion:"go1.19", Compiler:"gc", Platform:"linux/amd64"}
WARNING: version difference between client (1.21) and server (1.25) exceeds the supported minor version skew of +/-1
```

</details>

### Cloud provider

<details>
IBM Kubernetes Services (IKS)
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
NAME="Ubuntu"
VERSION="18.04.6 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.6 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic
$ uname -a
Linux test-cc38euv20td4f4nskdjg-standard125-default-0000033e 4.15.0-191-generic #202-Ubuntu SMP Thu Aug 4 01:49:29 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>
Containerd
</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by pacoxu**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:

https://github.com/kubernetes/kubernetes/blob/bcea98234f0fdc5ce4e976758977eec72b13bfd4/pkg/kubelet/images/helpers.go#L46-L51

The current rate limit is just an API qps/burst to calling the pull image request.
It doesn't wait for the image pulling done.
- as a result, if the qps and burst are 5, there can be more than 5 image pulling at the same time. 
- in the first second, start pulling 5 images
- in the second second, start pulling another 5 images

#### Which issue(s) this PR fixes:

Fixes #112044

#### Special notes for your reviewer:
The limit should wait until an image was pulled and start a new one

#### Does this PR introduce a user-facing change?

```release-note
None
```

### CODE REVIEW

File: pkg/kubelet/images/helpers.go [hunk 1]
```diff
@@ -33,19 +36,60 @@ func throttleImagePulling(imageService kubecontainer.ImageService, qps float32,
...
 	}
 	return &throttledImageService{
-		ImageService: imageService,
-		limiter:      flowcontrol.NewTokenBucketRateLimiter(qps, burst),
+		ImageService:     imageService,
+		limiter:          flowcontrol.NewTokenBucketRateLimiter(qps, burst),
+		qps:              qps,
+		burst:            burst,
+		currentPullCount: 0,
+		lock:             sync.Mutex{},
```
Comment (yxxhero) [inline pkg/kubelet/images/helpers.go:44]: lock don't need be init.

**Comment by Joseph-Goergen** [comment]
Just wondering what else is needed to get this in? Also, would it be possible to get this backported to 1.25 if this gets approved/merged?

**Comment by pacoxu** [comment]
/test pull-kubernetes-e2e-kind

> Just wondering what else is needed to get this in? 

It seems that no sig-node primary reviewer/approvers have reviewed this.

> Also, would it be possible to get this backported to 1.25 if this gets approved/merged?
If this is merged as a bugfix, I think backport is needed.
However, it may be a design problem in my mind. The implementation now is consistent with the qps/burst‘s literal meaning.

**Comment by pacoxu** [comment]
/close
see https://github.com/kubernetes/enhancements/issues/3673

**Pull request closed without merging**